### ISSUE TITLE
[4.1.0] Should orted call opal_set_using_threads()?

### ISSUE THREAD

**Issue by wzamazon** [open-mpi/ompi#8360]
Thank you for taking the time to submit an issue!

## Background information

orted in OMPI 4.1.0 randomly hit the following ```double free``` error:

```
*** Error in `/opt/amazon/openmpi/bin/orted': double free or corruption (fasttop): 0x000000000177bcf0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x816e9)[0x146811e9b6e9]
/opt/amazon/openmpi/lib64/libopen-rte.so.40(+0x3f3a7)[0x1468134883a7]
/opt/amazon/openmpi/lib64/libopen-rte.so.40(orte_rml_base_process_msg+0x40b)[0x1468134c606b]
/opt/amazon/openmpi/lib64/libopen-pal.so.40(opal_libevent2022_event_base_loop+0xd99)[0x1468131b8459]
/opt/amazon/openmpi/lib64/libopen-rte.so.40(orte_daemon+0x11e6)[0x146813481716]
/opt/amazon/openmpi/bin/orted[0x40077b]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x146811e3c575]
/opt/amazon/openmpi/bin/orted[0x4007b7]
```
I think the reason is  orted did not call ```opal_set_using_threads()``` to enable OPAL's multi thread support during initialization, even though it is using multiple threads.

However, I wonder why we did not see this error for open mpi 4.0.5, in which orted did not call ```opal_set_using_threads``` either.

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

4.1.0

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

built with the following configure options:
```
 '--build=x86_64-redhat-linux' '--host=x86_64-redhat-linux' '--target=x86_64-amazon-linux-gnu' '--program-prefix=' '--prefix=/opt/amazon/openmpi' '--exec-prefix=/opt/amazon/openmpi' '--bindir=/opt/amazon/openmpi/bin' '--sbindir=/opt/amazon/openmpi/sbin' '--sysconfdir=/opt/amazon/openmpi/etc' '--datadir=/opt/amazon/openmpi/share' '--includedir=/opt/amazon/openmpi/include' '--libdir=/opt/amazon/openmpi/lib64' '--libexecdir=/opt/amazon/openmpi/libexec' '--localstatedir=/var' '--sharedstatedir=/var/lib' '--mandir=/opt/amazon/openmpi/share/man' '--infodir=/usr/share/info' '--with-sge' '--without-verbs' '--with-libfabric=/opt/amazon/efa' 'LDFLAGS=-Wl,--build-id -Wl,--enable-new-dtags'
```

### Please describe the system on which you are running

* Operating system/version:  Amazon Linux 1
* Computer hardware: intel
* Network type: EFA

-----------------------------

## Details of the problem

When testing OMPI 4.1.0, I randomly hit the following ```double free``` error with ```orted```

command:
```
mpirun xxxx
```
Error message
```
*** Error in `/opt/amazon/openmpi/bin/orted': double free or corruption (fasttop): 0x000000000177bcf0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x816e9)[0x146811e9b6e9]
/opt/amazon/openmpi/lib64/libopen-rte.so.40(+0x3f3a7)[0x1468134883a7]
/opt/amazon/openmpi/lib64/libopen-rte.so.40(orte_rml_base_process_msg+0x40b)[0x1468134c606b]
/opt/amazon/openmpi/lib64/libopen-pal.so.40(opal_libevent2022_event_base_loop+0xd99)[0x1468131b8459]
/opt/amazon/openmpi/lib64/libopen-rte.so.40(orte_daemon+0x11e6)[0x146813481716]
/opt/amazon/openmpi/bin/orted[0x40077b]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x146811e3c575]
/opt/amazon/openmpi/bin/orted[0x4007b7]
```
I did some debug, and confirmed this is indeed a double free situation. Specifically, orted main thread and even processing thread are both calling ```OPAL_RLEASE``` to release a ```datacaddy_t``` object:

the even processing thread called ```relcbfunc``` (orted/pmix/pmix_server.c:583) to release 
the main thread called ```pmix_server_dmdx_resp()``` (orted/pmix/pmix_server.c:656) to release

I later found out that the reason for double free to happen is because OPAL's multi thread support is not turned on by default. Application need to call ```opal_set_using_threads(true)``` to turn it on, which orted did not do.

I tried to add call to ```opal_set_using_threads``` in orte_main.c, which seemed to solve this issue. (Because of the randomness of the issue, I cannot be 100% sure. By "seemed to",  I mean the issue did not show when running a testing program in a loop of 1000 times).

However, I wonder why we did not see this with open mpi 4.0.5? From what I can tell, orted in open mpi 4.0.5 did not call ```opal_set_using_threads``` either.

**Comment by rhc54** [comment]
Yes, it should call that fn - should do it in v4.0.x too.

**Comment by wzamazon** [comment]
@rhc54  Thank you for your reply!

I have one more question: I did some test with open mpi 4.0.5, and noticed for 4.0.5 the function ```pmix_server_dmdx_resp``` was not called, which should be the reason this bug does not appear for 4.05. 

I wonder why the behavior is different between 4.0.5 and 4.1.0?

**Comment by rhc54** [comment]
I suspect there was a change in the default setting controlling whether or not OMPI executes a full data exchange during MPI_Init. Sounds like OMPI is doing the full exchange in 4.0.x, but it is not doing so in 4.1.x. As a result, OMPI needs to execute a direct modex request for connection info in v4.1, which is why it would trip the bug.

**Comment by wzamazon** [comment]
@rhc54 Thank you so much!

I opened PR https://github.com/open-mpi/ompi/pull/8362, would you please take a look?

**Comment by rhc54** [comment]
Looked fine - thx!

**Comment by wzamazon** [comment]
PR https://github.com/open-mpi/ompi/pull/8362 was merged to 4.1.x

I opened another PR to back port it to 4.0.x:  https://github.com/open-mpi/ompi/pull/8366

**Comment by wzamazon** [comment]
both PR merged. closing ...
